### ISSUE TITLE
External token (not issued by Keycloak) cannot be validated in token exchange flow in case user info check is disabled

### ISSUE THREAD

**Issue by cvetkovv** [keycloak/keycloak#33332]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

token-exchange

### Describe the bug

When exchanging an access token issued by an external provider (e.g. Microsoft configured as OpenID Connect v1.0) and user info check disabled, keycloak is returning error `invalid_token` because the token doesn't contain claim `typ`. The 
`typ` claim is optional and some providers (e.g. Microsoft) are not setting it. The regression was introduced with https://github.com/keycloak/keycloak/pull/28866

Suggested change is to validate the `typ` claim only if it is available and make it required only in `KeycloakOIDCIdentityProvider` 

in `OIDCIdentityProvider` 
```
protected boolean isTokenTypeSupported(JsonWebToken parsedToken) {
    String type = parsedToken.getType();
    return Objects.isNull(type) || SUPPORTED_TOKEN_TYPES.contains(type);
}
```

and in `KeycloakOIDCIdentityProvider` 
```
protected boolean isTokenTypeSupported(JsonWebToken parsedToken) {
    String type = parsedToken.getType();
    if (Objects.isNull(type)) {
        return false;
    }
    return super.isTokenTypeSupported(parsedToken);
}
```

### Version

25.0.6

### Regression

- [X] The issue is a regression

### Expected behavior

Access token can be validated in case it doesn't contain `typ` claim.

### Actual behavior

External tokens that do not contain `typ` claim and user info check disabled cannot be exchanged to internal because of error `invalid_token`

### How to Reproduce?

Configure as `OpenID Connect v1.0` identity provider that doesn't set `typ` claim (e.g. Microsoft), disable user info check, and try to exchange external to internal token with grant_type = `urn:ietf:params:oauth:grant-type:token-exchange` and subject_token_type=`urn:ietf:params:oauth:token-type:jwt`.

**Comment by infl00p** [comment]
I recently upgraded to KC25.0.6 and I am affected by this issue. I left a comment in the PR #33344  to backport to KC25 since it's a regression and there are no workarounds.

**Comment by DenEwout** [comment]
Added a thumbs up for priority. 

- https://github.com/keycloak/keycloak/pull/33344: This is not a good fix for the problem because there are IDP's out there that just follow https://datatracker.ietf.org/doc/html/rfc7519#section-5.1 and put "JWT" as the "typ" claim value.
- https://github.com/keycloak/keycloak/pull/35075 would be more desirable.

**Comment by mbobzin** [comment]
We have a similar issue with Okta tokens, they don't contain the typ claim. [#35075](https://github.com/keycloak/keycloak/pull/35075) would solve our problem.

**Comment by rmartinc** [comment]
@pedroigor @keycloak/core-iam in general, I see this issue is triggered by the changes in this commit https://github.com/keycloak/keycloak/commit/0d60e5802915118f9c8a1a42382efcdb4af7ec3d. I suppose you are trying to limit the valid token using the `typ` claim. The problem is that there are a lot of external providers that are not filling this claim (for example google or microsoft). I don't know if this check should be moved to the specific `KeycloakOIDCIdentityProvider` or at least to provide a config option to allow disabling this check. I see the triggering issue was a security issue so we have to manage this with caution.

**Comment by pedroigor** [comment]
Yeah, that fix you mentioned was related to a security issue and I'm not sure if we can disallow this check without sacrifice the overall security of exchanges. Perhaps we can allow providing a list of token types via configuration. We need a bit more discussion to sort out how to properly address this one.

**Comment by pedroigor** [comment]
I see there is a PR for disabling checks https://github.com/keycloak/keycloak/pull/35075.

**Comment by rmartinc** [comment]
@pedroigor Yes, there are two PRs (one with a config option and another one that allows empty without any conf, I think that the config option solution is better). Don't know, you know the details more than me. IMHO the config option is OK, because it's secure by default, although adding the option for social providers that we already know do not add the `typ`, like google, is also a bit of a nonsense.

**Comment by infl00p** [comment]
> [@pedroigor](https://github.com/pedroigor) @keycloak/core-iam in general, I see this issue is triggered by the changes in this commit [0d60e58](https://github.com/keycloak/keycloak/commit/0d60e5802915118f9c8a1a42382efcdb4af7ec3d). I suppose you are trying to limit the valid token using the `typ` claim. The problem is that there are a lot of external providers that are not filling this claim (for example google or microsoft). I don't know if this check should be moved to the specific `KeycloakOIDCIdentityProvider` or at least to provide a config option to allow disabling this check. **I see the triggering issue was a security issue so we have to manage this with caution**.

Hello @rmartinc ,
Is this the relevant security notice https://github.com/keycloak/keycloak/security/advisories/GHSA-7fpj-9hr8-28vh ?
If correct, it seems to be missing from the changes of the 24.0.3 and 25.0.0 releases, or the Migration Changes docs.

Since this issue has the **help wanted** label, it is useful for the community to know the true reason for the https://github.com/keycloak/keycloak/pull/28866 and  https://github.com/keycloak/keycloak/commit/e3edf768678572d0302d83f6a4d477ddf97ddabb that  caused the problem.

**Comment by ahus1** [comment]
@rmartinc - it seems that I missed your recent comments here as I was only looking at he pull request. Sorry if I merged this too early.
